File: frontend/src/modules/crm/clientes/domain/IClienteRepository.ts
import type { Cliente } from './Cliente'
import type { ClienteFormValues } from '../application/clienteSchema'

/**
 * Repository contract for Cliente.
 * Story 2.1 introduced `getAll` + `getById`. Story 2.3 adds `create`.
 * Story 2.4 adds `update`.
 */
export interface IClienteRepository {
  getAll(signal?: AbortSignal): Promise<Cliente[]>
  getById(id: string, signal?: AbortSignal): Promise<Cliente>
  create(payload: ClienteFormValues, signal?: AbortSignal): Promise<Cliente>
  update(id: string, payload: ClienteFormValues, signal?: AbortSignal): Promise<Cliente>
}

File: backend/src/SiesaAgents.Domain/Repositories/IClienteRepository.cs
using SiesaAgents.Domain.Entities;

namespace SiesaAgents.Domain.Repositories;

/// <summary>
/// Repository contract for `clientes`. Read-only surface established in Story
/// 2.1/2.2; mutation methods (Create) added by Story 2.3. Update/Delete are
/// added by Stories 2.4/2.5.
/// </summary>
public interface IClienteRepository
{
    Task<IReadOnlyList<ClienteEntity>> GetAllAsync(string? searchTerm, CancellationToken ct);

    Task<ClienteEntity?> GetByIdAsync(Guid id, CancellationToken ct);

    Task AddAsync(ClienteEntity cliente, CancellationToken ct);

    Task<ClienteEntity?> UpdateAsync(ClienteEntity cliente, CancellationToken ct);

    /// <summary>
    /// Deletes the client with the given <paramref name="id"/>. Returns
    /// <c>false</c> (no exception) if no client with that id exists — the
    /// 404 case. Associated `contactos` are orphaned (`cliente_id = NULL`)
    /// by the database's `ON DELETE SET NULL` FK behavior (Story 2.5, R2),
    /// never cascade-deleted or nulled out via application code.
    /// </summary>
    Task<bool> DeleteAsync(Guid id, CancellationToken ct);

    /// <summary>
    /// Counts `contactos` associated with the given client, used only to
    /// decide which of the two delete-success toast variants applies
    /// (Story 2.5, AC #2 vs #3) — a narrow, story-scoped read, not a
    /// replacement for the full `IContactoRepository` Epic 3 will introduce.
    /// </summary>
    Task<int> CountContactosByClienteIdAsync(Guid clienteId, CancellationToken ct);
}
